stan/math/rev/functor/solve_newton.hpp

```diff
@@ -252,9 +348,9 @@ Eigen::Matrix<scalar_type_t<T2>, Eigen::Dynamic, 1> algebra_solver_newton(
     const double scaling_step_size = 1e-3,
     const double function_tolerance = 1e-6,
     const long int max_num_steps = 200) {  // NOLINT(runtime/int)
-  return algebra_solver_newton_impl(algebra_solver_adapter<F>(f), x, msgs,
-                                    scaling_step_size, function_tolerance,
-                                    max_num_steps, y, dat, dat_int);
+  return solve_newton_impl(algebra_solver_adapter<F>(f), x, msgs,
```

Comment:
The way we generic variadics in the compiler (e.g. for reduce_sum, ODES, DAEs) we place the `pstream__` argument after the required args and before the variadics. This `algebra_solver_adapter` seems to re-order them so that this isn't necessary, but it would actually be _more_ work if this didn't follow that pattern.

I would like in the future to do an internal change so that we don't have any of this kind of reordering (and move it all into stan_math) so that the pstream argument is always last, but for now, would you mind changing this?

Reply:
I should add that if you delete the wrapping in `algebra_solver_adapter`, adding these to the compiler looks like this:
```ocaml
  add_variadic_fn "solve_newton" ~return_type:UnsizedType.UVector
    ~control_args:[UnsizedType.(AutoDiffable, UVector)]
    ~required_fn_rt:UnsizedType.UVector
    ~required_fn_args:[UnsizedType.(AutoDiffable, UVector)]
    ()
```

That's all that would be required for each signature. Otherwise we will need to special-case the code generation

Reply:
My understanding is that a function defined in Stan is translated into a functor with an operator which takes `msgs` as its last argument. In `stan-math` (i.e. at the C++ level), the variadic arguments always need to be passed last. So some reordering needs to happen. Or can I write a statement like the following:
```
return f(x, args..., msgs)
```
If the above is valid, we can remove `algebra_solver_adapter` all together.

Reply:
`msg` is the final argument for non-variadic versions. We have special code gen for all other variadic functions, so for now I think we should for these. 

If we could have stan_math always pass it at the end that would be nice, but that would be a separate PR imo

Reply:
Compare to `dae`, which has the `msg` in the middle before `args...`: 

https://github.com/stan-dev/math/blob/224e5812c0c4885f5420caa72275f64f0b2568ed/stan/math/rev/functor/dae_system.hpp#L149

To be perfectly clear, I _wish_ it was always the case that `msgs` was the last argument, but at the moment it isn't for any of the functions accepted by variadic functions, so we should be consistent.

Reply:
> msg is the final argument for non-variadic versions. 

The function, `f`, which specifies the algebraic system and is defined by the user in the `functions` block is a non-variadic function. Once translated into C++, it might look as follows:
```
struct simple_eq_functor {
  template <typename T1, typename T2, typename T3, typename T4>
  inline Eigen::Matrix<stan::return_type_t<T1, T2>, Eigen::Dynamic, 1>
  operator()(const T1& x, const T2& y, const T3& dat, const T4& dat_int,
             std::ostream* pstream__) const {
    Eigen::Matrix<stan::return_type_t<T1, T2>, Eigen::Dynamic, 1> z(2);
    z(0) = x(0) - y(0) * y(1);
    z(1) = x(1) - y(2);
    return z;
  }
};
```
The user doesn't specify the `msgs` argument, this simply gets tacked at the end during translation. So we need to handle a function which takes msgs as its last argument. (unless something has changed regarding how user-defined functions are translated)

This is also how the ode solvers are tested. If I look at the functor used for the unit tests, e.g. for the [harmonic oscillator](https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/functor/harmonic_oscillator.hpp#:~:text=%7D%3B-,struct%20harm_osc_ode_data_fun%20%7B,%7D,-%3B):
```
struct harm_osc_ode_fun {
  template <typename T0, typename T1, typename T2>
  inline std::vector<stan::return_type_t<T1, T2>>
  // initial time
  // initial positions
  // parameters
  // double data
  // integer data
  operator()(const T0& t_in, const std::vector<T1>& y_in,
             const std::vector<T2>& theta, const std::vector<double>& x,
             const std::vector<int>& x_int, std::ostream* msgs) const {
    if (y_in.size() != 2)
      throw std::domain_error(
          "this function was called with inconsistent state");

    std::vector<stan::return_type_t<T1, T2>> res;
    res.push_back(y_in.at(1));
    res.push_back(-y_in.at(0) - theta.at(0) * y_in.at(1));

    return res;
  }
};
```

I'm not sure what's going on with dae. The [functor used in the unit test](https://github.com/stan-dev/math/blob/0a53c7d78cf01128dca6c6c46f137cd84aebb901/test/unit/math/rev/functor/idas_system_test.cpp) indeed has `msgs` in the middle:
```
struct chemical_kinetics {
  template <typename T0, typename Tyy, typename Typ, typename Tpar>
  inline Eigen::Matrix<stan::return_type_t<Tyy, Typ, Tpar>, -1, 1> operator()(
      const T0& t_in, const Eigen::Matrix<Tyy, -1, 1>& yy,
      const Eigen::Matrix<Typ, -1, 1>& yp, std::ostream* msgs,
      const std::vector<Tpar>& theta, const std::vector<double>& x_r,
      const std::vector<int>& x_i) const {
    if (yy.size() != 3 || yp.size() != 3)
      throw std::domain_error(
          "this function was called with inconsistent state");

    Eigen::Matrix<stan::return_type_t<Tyy, Typ, Tpar>, -1, 1> res(3);

    auto yy1 = yy(0);
    auto yy2 = yy(1);
    auto yy3 = yy(2);

    auto yp1 = yp(0);
    auto yp2 = yp(1);
    // auto yp3 = yp.at(2);

    auto p1 = theta.at(0);
    auto p2 = theta.at(1);
    auto p3 = theta.at(2);

    res[0] = yp1 + p1 * yy1 - p2 * yy2 * yy3;
    res[1] = yp2 - p1 * yy1 + p2 * yy2 * yy3 + p3 * yy2 * yy2;
    res[2] = yy1 + yy2 + yy3 - 1.0;

    // jacobian respect to yy

    return res;
  }
};
```
But I don't think that's the type of structure you get when Stan translates a user-defined function. Has the DAE been exposed to the Stan language and tested?

Reply:
Yes, and we generate unique functors for the variadic functions each time. So a ODE will have foo_functor_ and (currently called) foo_odefunctor_ if you inspect a generated hpp file. DAES have a foo_daefunctor_. All these do is call the original with the arguments re-ordered

I was previously told this was a necessary technical requirement, but obviously there is something we can (and in the future should) be doing in the math library to allow us to just use the _functor_ version everywhere and get rid of these extra bespoke functor declarations

Reply:
Ok, great. This now makes sense to me. So the higher-order functions with variadic arguments should target functions with `msgs` appearing in the middle (e.g foo_odefunctor, foor_daefunctor).

I now see your point and I'll make the requested changes.

Reply:
Thanks! Sorry for not being clear originally